### PR TITLE
MINIFI-153 - Adding proxy settings to remote process group

### PR DESCRIPTION
**Pull request by brosander**
Thank you for submitting a contribution to Apache NiFi - MiNiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with MINIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi-minifi folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] - N/A - If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [x] - N/A - If applicable, have you updated the LICENSE file, including the main LICENSE file under minifi-assembly?
- [x] - N/A - If applicable, have you updated the NOTICE file, including the main NOTICE file found under minifi-assembly?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by brosander** [comment]
Tested with [MINIFI-153 branch of my testkit.](https://github.com/brosander/minifi-testkit/tree/MINIFI-153)

**Comment by JPercivall** [comment]
Reviewing

File: minifi-commons/minifi-commons-schema/src/main/java/org/apache/nifi/minifi/commons/schema/RemoteProcessGroupSchema.java [hunk 1]
```diff
@@ -91,6 +98,10 @@ public RemoteProcessGroupSchema(Map map) {
         result.put(TIMEOUT_KEY, timeout);
         result.put(YIELD_PERIOD_KEY, yieldPeriod);
         result.put(TRANSPORT_PROTOCOL_KEY, transportProtocol);
+        result.put(PROXY_HOST_KEY, proxyHost);
+        result.put(PROXY_PORT_KEY, proxyPort == null ? "" : proxyPort);
```
Comment (JPercivall) [inline minifi-commons/minifi-commons-schema/src/main/java/org/apache/nifi/minifi/commons/schema/RemoteProcessGroupSchema.java]: Why is Port the only one with a null check?

File: minifi-commons/minifi-commons-schema/src/main/java/org/apache/nifi/minifi/commons/schema/RemoteProcessGroupSchema.java [hunk 2]
```diff
@@ -78,9 +80,14 @@ public RemoteProcessGroupSchema(Map map) {
         yieldPeriod = getOptionalKeyAsType(map, YIELD_PERIOD_KEY, String.class, wrapperName, DEFAULT_YIELD_PERIOD);
         transportProtocol = getOptionalKeyAsType(map, TRANSPORT_PROTOCOL_KEY, String.class, wrapperName, DEFAULT_TRANSPORT_PROTOCOL);
 
-        if (!transportProtocolOptions.valid(transportProtocol)){
+        if (!TransportProtocolOptions.valid(transportProtocol)){
             addValidationIssue(TRANSPORT_PROTOCOL_KEY, wrapperName, "it must be either 'RAW' or 'HTTP' but is '" + transportProtocol + "'");
         }
+
+        proxyHost = getOptionalKeyAsType(map, PROXY_HOST_KEY, String.class, wrapperName, DEFAULT_PROXY_HOST);
```
Comment (JPercivall) [inline minifi-commons/minifi-commons-schema/src/main/java/org/apache/nifi/minifi/commons/schema/RemoteProcessGroupSchema.java]: I believe we should add some more validation checks here. Such as if the host or port is set then the other must be. Also if username is set then password is set. Thoughts?

File: minifi-commons/minifi-commons-schema/src/main/java/org/apache/nifi/minifi/commons/schema/v2/RemoteProcessGroupSchemaV2.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    private String url;
+    private List<RemoteInputPortSchema> inputPorts;
+
+    private String comment = DEFAULT_COMMENT;
+    private String timeout = DEFAULT_TIMEOUT;
+    private String yieldPeriod = DEFAULT_YIELD_PERIOD;
+    private String transportProtocol = DEFAULT_TRANSPORT_PROTOCOL;
+
+    public RemoteProcessGroupSchemaV2(Map map) {
+        idAndName = new BaseSchemaWithIdAndName(map, "RemoteProcessGroup(id: {id}, name: {name})");
```
Comment (brosander) [inline minifi-commons/minifi-commons-schema/src/main/java/org/apache/nifi/minifi/commons/schema/v2/RemoteProcessGroupSchemaV2.java]: If you extend it, you're forced to implement toMap() when older versions don't really need to do that or even implement the WritableSchema interface.  Switching this to composition let me delete more code.

File: minifi-commons/minifi-commons-schema/src/main/java/org/apache/nifi/minifi/commons/schema/RemoteProcessGroupSchema.java (hunk 2 above)
Comment (brosander) [inline minifi-commons/minifi-commons-schema/src/main/java/org/apache/nifi/minifi/commons/schema/RemoteProcessGroupSchema.java]: +1

**Comment by JPercivall** [comment]
Hey @brosander, I just verified with secure and in-secure proxies hitting multiple different logical networks. All worked as expected.

One more thing that should probably be added is a validation such that if RAW is set then the proxy config should not be. (Since they are mutually exclusive)

Aside from that, I am a +1

**Comment by brosander** [comment]
@JPercivall added validation that if proxy host is set, transport protocol should be http

**Comment by JPercivall** [comment]
+1

Visually verified code and did a contrib-check build. Ran multiple different flows which touched a regular NiFi instance (RAW and HTTP) and utilized a secure and unsecure proxy to S2S to instances that were on a different logical network. Everything worked as expected.

Thanks @brosander, I will merge it in.

**Pull request closed without merging**